Graphiql App

ramzes-84 opened the pull request:
1. Task: https://github.com/rolling-scopes-school/tasks/blob/master/react/modules/graphiql.md
2. Screenshot: 
![image](https://github.com/ramzes-84/graphiql-app/assets/46446962/01ddd579-2654-4903-8503-381bf6fae7fd)
3. Deploy: https://graphiql-app-gamma.vercel.app
4. Done 8.01.2024 / deadline 8.01.2024
5. Video intro: https://youtu.be/ev6vyyhjl6g
6. Score: 350 / 350
7. For running app locally in dev mode, .env file required. Please contact us if it needed.

### Welcome route - max 50 points

- [x] The welcome page should contain general information about the developers, project, and course **10 points**
- [x] In the upper right corner there are 2 buttons: Sign In and Sign Up **10 points**
- [x] If the login token is valid and unexpired, the Sign In and Sign Up buttons are replaced with the "Main Page" button **10 points**
- [x] When the token expires - the user should be redirected to the "Welcome page" automatically **10 points**
- [x] Pressing the Sign In / Sign up button redirects a user to the route with the Sign In / Sign up form **10 points**

### Sign In / Sign Up - max 50 points

- [x] Buttons for Sign In / Sign Up / Sign Out are everywhere where they should be **10 points**
- [x] Client-side validation is implemented **20 points**
- [x] Upon successful login, the user is redirected to the Main page **10 points**
- [x] If the user is already logged in and tries to reach these routes, they should be redirected to the Main page **10 points**

### GraphiQL route - max 200 points

- [x] Functional editor enabling query editing and prettifying **60 points**
- [x] Operational documentation explorer, visible _only_ upon successful SDL request **50 points**
- [x] Variables section that can shown or hidden, specified variables are sent to the server **30 points**
- [x] Header section that can be shown or hidden, user-added headers are sent to the server **20 points**
- [x] Response section with an editor in read-only as a JSON viewer **40 points**

### General requirements - max 50 points

- [x] Localization **30 points**
- [x] Sticky header **10 points**
- [x] Errors from API side are displayed in the user friendly format **10 points**

### Penalties are absent

Commit:
init: create next app

Commit:
feat: add prettier, format files

Commit:
feat: add eslint & ts rules

Commit:
fead: add husky, add lint staged & configuring

Commit:
docs: add instruction for linux users how to fix husky

Commit:
fix: temporary remove pre-push hook

Commit:
feat: add jest, add hook

Commit:
test: add first test

Commit:
feat: add favicon, remove sample code

Commit:
fix: remove sample code

Commit:
feat: add header & footer to layout, add tests coverage

Commit:
feat: add 404 page, add test

MKsenni committed:
feat: add change of header by scrolling

MKsenni committed:
feat: add link to welcome page and icon for sign out

MKsenni committed:
fix: add signin icon

Commit:
feat: add error boundary, add tests, exlude layout.tsx
from coverage

Commit:
fix: remove unused async

MKsenni committed:
fix: deleted tag

MKsenni committed:
Merge pull request #5 from ramzes-84/merk_branch

ramzes-84 committed:
fix: replace Main page with Welcome page

ramzes-84 committed:
Merge pull request #6 from ramzes-84/error-bound

Add error boundary, exlude layout.tsx from coverage

Commit:
test: add test for layout

ramzes-84 committed:
Merge pull request #7 from ramzes-84/develop

PR

Commit:
config: add layout.tsx to conerage, set coverage to 80

ramzes-84 committed:
Merge pull request #8 from ramzes-84/some-tests

Add test for layout, add layout.tsx to coverage, set coverage to 80

Commit:
feat: add global provider wrapper, add lang controls to header

Commit:
test: add test for lang btns

Commit:
feat: add l18n to app

Commit:
refactor: remove data-testid

ramzes-84 committed:
Merge pull request #9 from ramzes-84/en-ru

Add context provider, add l18n

Commit:
feat: add footer

Commit:
feat: add Welcome page

ramzes-84 committed:
Merge pull request #10 from ramzes-84/footer

Add footer

lu7623 committed:
feat: add user auth

Commit:
feat: add personal info to dict

Commit:
feat: add course logo

lu7623 committed:
feat: add translation, add some tests

lu7623 committed:
feat: add firebase adapter

ramzes-84 committed:
Merge pull request #11 from ramzes-84/welcome

Welcome page info

lu7623 committed:
feat: add email-password login

lu7623 committed:
feat: add translate forms

lu7623 committed:
feat: add some tests

MKsenni committed:
feat: add editor view

MKsenni committed:
Merge remote-tracking branch 'origin' into merk_branch_2

MKsenni committed:
refactor: add dictionary for editor

MKsenni committed:
fix: some fix

MKsenni committed:
refactor: refactor jest settings

MKsenni committed:
fix: fix jest mistake

MKsenni committed:
fix: fix jest mistake

MKsenni committed:
fix: fix jest mistake

MKsenni committed:
fix: some fix

MKsenni committed:
fix: some fix

MKsenni committed:
fix: bug fix in test, change function test coverage

lu7623 committed:
feat: form validation

lu7623 committed:
feat: user-friendly errors

MKsenni committed:
feat: add formatting function

MKsenni committed:
fix: fix build error and change editor component

MKsenni committed:
feat: add tests

MKsenni committed:
Merge pull request #12 from ramzes-84/merk_branch_2

lu7623 committed:
feat: save language to ls

lu7623 committed:
Merge branch 'develop' into user-auth

lu7623 committed:
fix: fix test

lu7623 committed:
feat: disable buttons

lu7623 committed:
fix: build fix

lu7623 committed:
Merge pull request #13 from ramzes-84/user-auth

User auth with Next-auth & Firebase

lu7623 committed:
feat: set expitation time

lu7623 committed:
feat: separate sign in and sign up pages

MKsenni committed:
refactor: add service-account in gitignor

MKsenni committed:
Merge remote-tracking branch 'origin' into merk_branch_3

MKsenni committed:
feat: add example of request function

MKsenni committed:
feat: component viewer made

MKsenni committed:
refactor: made viewer readonly

MKsenni committed:
feat: add viewer in json format

MKsenni committed:
feat: add tests

ramzes-84 committed:
refactor: welcome page & tests for it

MKsenni committed:
fix: fix of comments

ramzes-84 committed:
feat: add btns, fix tests

ramzes-84 committed:
Merge branch 'welcome-btns' into merk_branch_3

ramzes-84 committed:
Merge pull request #15 from ramzes-84/merk_branch_3

Merk branch 3

ramzes-84 committed:
test: refactor welcome page tests

MKsenni committed:
fix: fix comments

MKsenni committed:
Merge branch 'merk_branch_3' of https://github.com/ramzes-84/graphiql-app into merk_branch_3

MKsenni committed:
Merge pull request #14 from ramzes-84/merk_branch_3

ramzes-84 committed:
Merge pull request #17 from ramzes-84/develop

Merge

lu7623 committed:
feat: merge dvelop into user-token

lu7623 committed:
feat: add tests

lu7623 committed:
fix: fix welcome page btns

ramzes-84 committed:
Merge pull request #16 from ramzes-84/welcome-btns

Buttons on Welcome Page for auth & not auth users

lu7623 committed:
fix: some fix

lu7623 committed:
fix: fix coverage

ramzes-84 committed:
feat: add service chooser

ramzes-84 committed:
Merge pull request #20 from ramzes-84/user-token

fix: fix coverage

lu7623 committed:
Merge pull request #18 from ramzes-84/user-token

User token

Commit:
refactor: remove if from server-chooser

Commit:
fix: add translation

MKsenni committed:
feat: add mode for highlighting

MKsenni committed:
fix: some fix in formatecode function

MKsenni committed:
fix: some fix of tests

ramzes-84 committed:
fix: fix issues

Commit:
feat: implement Lu's suggestion

Commit:
fix: styles

ramzes-84 committed:
Merge pull request #21 from ramzes-84/service-chooser

Server chooser

lu7623 committed:
feat: add adaptive

lu7623 committed:
feat: add styles to error pages

lu7623 committed:
fix: fix a line

lu7623 committed:
Merge pull request #22 from ramzes-84/adaptive

Adaptive up to 320px

MKsenni committed:
Merge remote-tracking branch 'origin' into prettifying

MKsenni committed:
feat: add change in formatCode function

MKsenni committed:
fix: add node env in script test

MKsenni committed:
fix: add node env test in prepush

lu7623 committed:
feat: add lower panel ui

lu7623 committed:
feat: modify request with variables

lu7623 committed:
feat: add some tests

lu7623 committed:
feat: add tests

lu7623 committed:
feat: add request context

lu7623 committed:
fix: fix tests

lu7623 committed:
feat: add reducer

lu7623 committed:
fix: fix tests

lu7623 committed:
fix: minor tests fix

lu7623 committed:
fix: try another spinner pic

MKsenni committed:
fix: attempt to correct error of env and test

lu7623 committed:
Merge pull request #23 from ramzes-84/variables

Variables

MKsenni committed:
Merge remote-tracking branch 'origin' into prettifying

lu7623 committed:
feat: add headers to request

lu7623 committed:
feat: add tests

lu7623 committed:
fix: minor test fix

lu7623 committed:
feat: add handling some errors

lu7623 committed:
Merge pull request #24 from ramzes-84/headers

Headers

MKsenni committed:
Merge remote-tracking branch 'origin' into prettifying

MKsenni committed:
fix: fix of incorrect enter

MKsenni committed:
fix: fix enter after comma, fix one to two spaces for step, fix query in []

MKsenni committed:
Merge pull request #25 from ramzes-84/prettifying

lu7623 committed:
feat: add user-friendly messages

lu7623 committed:
fix: some fix

lu7623 committed:
fix: fix request

lu7623 committed:
feat: add tests

lu7623 committed:
fix: fix conflicts

lu7623 committed:
fix: some fix

lu7623 committed:
fix: fix test

lu7623 committed:
fix: fix styles

Commit:
feat: add docs explorer

lu7623 committed:
Merge pull request #27 from ramzes-84/handling-errors

Handling errors

Commit:
fix: fix writing new schema

Commit:
test: add tests

Commit:
test: add tests

ramzes-84 committed:
Merge branch 'develop' into explorer

MKsenni committed:
feat: realized server action for send request

Commit:
fix: remove unused async

ramzes-84 committed:
Merge pull request #28 from ramzes-84/explorer

Add docs explorer

lu7623 committed:
feat: add next middleware

MKsenni committed:
Merge remote-tracking branch 'origin' into merk_branch_4

MKsenni committed:
fix: fix some tests

MKsenni committed:
Merge pull request #31 from ramzes-84/merk_branch_4

lu7623 committed:
Merge pull request #30 from ramzes-84/middleware

Add next middleware

lu7623 committed:
feat: add try catch

lu7623 committed:
fix: fix tests

lu7623 committed:
fix: fix styles

lu7623 committed:
fix: some fix

lu7623 committed:
fix: get rid of all console errors in tests

lu7623 committed:
fix: delete unnecessary file

lu7623 committed:
feat: update README.md

lu7623 committed:
fix: update .gitignore

lu7623 committed:
fix: fix docs section logic

lu7623 committed:
Merge branch 'fetch-needs-try-catch' of https://github.com/ramzes-84/graphiql-app into fetch-needs-try-catch

lu7623 committed:
Merge pull request #32 from ramzes-84/fetch-needs-try-catch

Minor fixes

leryor commented on app/components/variables.tsx:
```diff
@@ -0,0 +1,31 @@
+      setText(newText);
+      dispatch({ type: "setVariables", payload: newText });
+    },
+    [dispatch]
```
I don't think `dispatch` function can change, so empty dependency array is enough

leryor commented on app/main/page.tsx:
```diff
@@ -0,0 +1,158 @@
+  const [error, setError] = useState("");
+  const [schemaErrors, setSchemaErrors] = useState("");
+  useEffect(() => {
+    async function getData() {
```
I believe we don't need async function here as we don't use await. So this wrap function is redundant.

leryor commented on app/page.tsx:
```diff
@@ -0,0 +1,102 @@
+
+export default function Home() {
+  const dict = useDict();
+  const [isOpen, setIsOpen] = useState<number | false>(false);
```
Don't understand why we need `number` here.

leryor commented on app/signup/page.tsx:
```diff
@@ -0,0 +1,58 @@
+    const email = data.email;
+    const password = data.password;
+
+    await createUserWithEmailAndPassword(auth, email, password)
```
I would suggest do not mix an async/await and callbacks. Choose one approach to avoid a confusion.

leryor commented on firebase.ts:
```diff
@@ -0,0 +1,24 @@
+  storageBucket: "graphql-app-a8e63.appspot.com",
+  messagingSenderId: "550201743702",
+  appId: "1:550201743702:web:a8d11be80c52c362f7def7",
+  measurementId: "G-9K4LFM0M96",
```
This is quite a serious security and configuration issue. All values above should be in env variables.

leryor commented on app/utils/request.ts:
```diff
@@ -0,0 +1,57 @@
+  headers?: string
+): Promise<FullSchema> {
+  try {
+    const res = await fetch(endpoint, {
```
Please don't mix async/await with callbacks.

leryor reviewed (approved):
@ramzes-84 @lu7623 @MKsenni 
Guys, this is a really brilliant job, keep it up! I'm proud of you!

lu7623 committed:
fix: fix firebase config

lu7623 committed:
fix: fix async/callback

lu7623 committed:
fix: minor fix

lu7623 committed:
Merge pull request #33 from ramzes-84/mentor-review-fixes

Mentor review fixes

dependabot[bot] committed:
build(deps): bump protobufjs and firebase-admin

Bumps [protobufjs](https://github.com/protobufjs/protobuf.js) to 7.2.6 and updates ancestor dependency [firebase-admin](https://github.com/firebase/firebase-admin-node). These dependencies need to be updated together.


Updates `protobufjs` from 7.2.4 to 7.2.6
- [Release notes](https://github.com/protobufjs/protobuf.js/releases)
- [Changelog](https://github.com/protobufjs/protobuf.js/blob/master/CHANGELOG.md)
- [Commits](https://github.com/protobufjs/protobuf.js/compare/protobufjs-v7.2.4...protobufjs-v7.2.6)

Updates `firebase-admin` from 11.11.1 to 12.0.0
- [Release notes](https://github.com/firebase/firebase-admin-node/releases)
- [Commits](https://github.com/firebase/firebase-admin-node/compare/v11.11.1...v12.0.0)

---
updated-dependencies:
- dependency-name: protobufjs
  dependency-type: indirect
- dependency-name: firebase-admin
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

ramzes-84 committed:
Merge pull request #34 from ramzes-84/dependabot/npm_and_yarn/multi-a1e966211b

build(deps): bump protobufjs and firebase-admin
